packages/plugins/response-cache/src/index.ts

```diff
@@ -0,0 +1,208 @@
+/* eslint-disable no-console */
+import { Plugin } from '@envelop/types';
+import LRU from 'tiny-lru';
+import { createHash } from 'crypto';
+import { DocumentNode, OperationDefinitionNode, FieldNode, SelectionNode, visit, parse, print } from 'graphql';
+
+type Listener = (typename: string, id?: string | number) => void;
+
+interface Controller {
+  purge(typename: string, id?: string | number): void;
+  ɵregister(listener: Listener): void;
+}
+
+interface Options {
+  max?: number;
+  ttl?: number;
+  controller?: Controller;
+}
+
+export function createController(): Controller {
+  let listener: (typename: string, id?: string | number) => void = () => {};
+
+  return {
+    purge(typename, id) {
+      listener(typename, id);
+    },
+    ɵregister(audience) {
+      listener = audience;
+    },
+  };
+}
+
+export const useResponseCache = ({ max = 1000, ttl = undefined, controller }: Options = {}): Plugin => {
+  if (controller) {
+    controller.ɵregister((typename, id) => {
+      purge(typeof id !== 'undefined' ? makeId(typename, id) : typename);
+    });
+  }
+
+  const cachedResponses = LRU<any>(max, ttl);
+
+  // TODO: when TTL or max size is reached, we need to clean the maps below
+  // TODO: Let's maybe use a different LRU cache (with a callback on delete action)
+  const entityToResponse = new Map<string, Set<string>>();
+  const responseToEnity = new Map<string, Set<string>>();
+
+  function purge(entity: string) {
+    if (entityToResponse.has(entity)) {
+      const responsesToRemove = entityToResponse.get(entity);
+
+      if (responsesToRemove) {
+        responsesToRemove.forEach(responseId => {
+          // get entities related to the response
+          if (responseToEnity.has(responseId)) {
+            responseToEnity.get(responseId)!.forEach(entityId => {
+              // remove the response mapping from the entity
+              entityToResponse.get(entityId)?.delete(responseId);
+            });
+            // remove all the entity mappings from the response
+            responseToEnity.delete(responseId);
+          }
+          // remove the response from the cache
+          cachedResponses.delete(responseId);
+        });
+      }
+    }
+  }
+
+  return {
+    onParse(ctx) {
+      ctx.setParseFn((source, options) => addTypenameToDocument(parse(source, options)));
+    },
+    onExecute(ctx) {
+      if (isMutation(ctx.args.document)) {
+        return {
+          onExecuteDone({ result }) {
+            const entitiesToRemove = new Set<string>();
+
+            collectEntity(result.data, (typename, id) => {
+              if (typeof id !== 'undefined') {
+                entitiesToRemove.add(makeId(typename, id));
+              }
+            });
+
+            entitiesToRemove.forEach(purge);
+          },
+        };
+      } else {
+        const operationId = createHash('md5').update(print(ctx.args.document)).digest('hex');
```

Comment:
how should caching per user session work?

Reply:
also sha might be better than md5

Reply:
Based on a value defined in a context (custom function to set this up based on req.sessionId or whatever)

Reply:
@n1ru4l I added the session thing